#####Melanocyte selective genes
######Update Aug 5 2021
#library
library(dplyr)
library(plyr)
### Load Data

proteintogene <- read.table("uniprot-filtered-organism__Homo+sapiens+(Human)+[9606]_.tab", header=T, sep="\t", quote="", comment.char = "")
head(proteintogene)

#Proteomics analysis of Skin
#https://www.nature.com/articles/s41467-020-19383-8
#Spatially and cell-type resolved quantitative proteomic atlas of healthy human skin
# Beatrice Dyring-Andersen, Marianne Bengtson Løvendorf, Fabian Coscia, Alberto Santos, Line Bruun Pilgaard Møller, Ana R. Colaço, Lili Niu, Michael Bzorek, Sophia Doll, Jørgen Lock Andersen, Rachael A. Clark, Lone Skov, Marcel B. M. Teunissen & Matthias Mann 
# Nature Communications volume 11, Article number: 5587 (2020) Cite this article

importskinatlas <- function(){
  skinatlas <- read.table("skinatlas_data_data.tsv", header=T, sep="\t", stringsAsFactors = F)
  skinatlas <- as.data.frame(t(skinatlas),stringsAsFactors = F)
  colnames(skinatlas)  <- as.character(skinatlas[1,])
  skinatlas <- skinatlas[-1,]
  skinatlas$ids <- row.names(skinatlas)
    proteintoGene <- merge(data.frame(ids=skinatlas$ids), proteintogene[,c(1:2,5)],by.x="ids", by.y="Entry", all.x=T )
      skinatlas <- join(skinatlas, proteintoGene)
      skinatlas$Entry.name <- gsub("_HUMAN", "", skinatlas$Entry.name)
      skinatlas[,1:13] <- as.data.frame(lapply(skinatlas[,1:13], as.numeric))
  return(skinatlas)
}

skinatlas <- importskinatlas()

### Functions
#From:Fantom5 database
#Function to select Melanocayte Specific genes from Fantom 5 Data
# 3 cell types in Fantom5  "Melanocyte.dark."  "Melanocyte.light." "Melanocyte."   
# Select Max of three vs max expression in any other cell type
MelanocyteSelective <- function(num=2){
  if(!exists("MeanCluster_Fantom_t")) {
    load("./Data/MembraneDataAll.RData")
  }
  df <- Fantom
  MelanocyteIndex <- grep("Melan", names(df))
    MelanocyteMax <- apply(df[,MelanocyteIndex], 1, max)
    OthersMac     <- apply(df[,c(-1, -MelanocyteIndex)], 1, max)
    OthersMac     <- apply(df[,c(-1, -MelanocyteIndex)], 1, function(x)(sort(x,TRUE)[num]))
  MelaSelFC <- ((MelanocyteMax+1)/(OthersMac+1))
  DF <- data.frame(df$X, MelanocyteMax, OthersMac,  MelaSelFC)
  DF <- DF[order(DF$MelaSelFC, decreasing = T),]
  return(DF)
}

###From GTEX
#Function to select Skin Specific genes from GTEX Data
# Select Max Exprs in Skin  vs max expression in any other cell type
SkinSelective <- function(brain=TRUE){
  if(!exists("Gtex")) {
    load("./Data/MembraneDataAll.RData")
  }
  
  dfSkin    <-data.frame(Gtex[,3:55])
  SkinIndex <- grep("Skin", names(dfSkin))
  BrainIndex <- grep("Brain|Nerve", names(dfSkin))
    SkinMax   <- apply(dfSkin[,SkinIndex], 1, max)
    OthersMax <- apply(dfSkin[,c(-1, -SkinIndex)], 1, max)
    if (brain==TRUE){OthersMax <- apply(dfSkin[,c(-1, -SkinIndex,-BrainIndex)], 1, max)}
    SkinSelFC <- ((SkinMax+1)/(OthersMax+1))
  DFSkin    <- data.frame(Gtex$Description, SkinMax, OthersMax,  SkinSelFC)
    DFSkin   <- DFSkin[order(DFSkin$SkinSelFC, decreasing = T),]
  return(DFSkin)
}

#Merge(Fantom5 data (MelanocyteSelective) and Gtex data (DFSKIN)
MelaSkinMerge <- function(){
  Temp1 <- DFMela
    Temp1$genename<- gsub("p[0-9]+@","", Temp1$df.X)
  Temp2 <- merge(Temp1, DFSkin, by.x="genename", by.y="Gtex.Description")
    Temp2 <- Temp2[order(Temp2$MelaSelFC, decreasing=T),]
  return(Temp2)
}

uniprot_mapping <- function(ids) {
  uri <- 'http://www.uniprot.org/uniprot/?query='
  idStr <- paste(ids, collapse="+or+")
  format <- '&format=tab'
  fullUri <- paste0(uri,idStr,format)
  dat <- read.delim(fullUri)
  dat
}

SkinAtlasExpress <- function(gene="TYRP1"){
  for (i in gene){
    #print(i)
    IndexList <- c(IndexList, grep(paste0("^",i,"$"), skinatlas$Entry.name))
  }
  for (i in gene){
    IndexList <- c(IndexList, grep(paste0("^",i,"$"), skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0(i,"$"),     skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0("^",i," "), skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0(" ",i," "), skinatlas$Gene.names))
  }
  IndexList <- unique(IndexList)
  output <- skinatlas[IndexList,]
   
  
  return(output)
}  

SkinAtlasExpress()


###Select for melocyte Specific genes MelFC is the melocyte enrichment over all other cells in Fantom
#### SKIN FC is the enrichment in GTEX skin over all other cells 
SelectMelocyteSkinSpecificGenes <- function(Melfc=2, SkinFC=1){
  require(data.table) ## 1.9.2
  MelaSkinSpecific_Output <- MelanocyteSelective()
  MelaSkinSpecific_Output <- MelaSkinSpecific_Output[MelaSkinSpecific_Output$MelanocyteMax>5,]
    MelaSkinSpecific_Output  <- MelaSkinSpecific_Output[MelaSkinSpecific_Output$MelaSelFC >Melfc,]
    MelaSkinSpecific_Output$Gene <- gsub("p[0-9]+@", "", MelaSkinSpecific_Output$df.X)
    group <- as.data.table(MelaSkinSpecific_Output)
  #If you want to keep all the entries corresponding to max values of pt within each group:
  #Select transcript for gene with the highest expression
    group <- unique(group[group[, .I[MelaSelFC == max(MelaSelFC)], by=Gene]$V1])
    
    DFSkin <- SkinSelective() #Gtex
    DFSkin <- DFSkin[DFSkin$SkinSelFC>SkinFC,]
    
    #Merge Melanocyte specific and Skin specific
    Output <- merge(group, DFSkin, by.x="Gene", by.y="Gtex.Description")
      Output <- unique(Output)
      Output <- Output[order(Output$MelaSelFC, decreasing = TRUE)]
      SkinProtien <- SkinAtlasExpress(gene = Output$Gene)

  return(list(Output, SkinProtien))
}


SelectMelocyteSkinSpecificGenes(Melfc = 2, SkinFC = 0)[[1]]

DFSkin <- SkinSelective() #Gtex
DFMela <- MelanocyteSelective(num=2)#Fantom5, Num = 2 is the second highest expression in non melanocytes
SelectMelocyteSkinSpecificGenes(Melfc=2, SkinFC=1.5)[1:100,1]

#Select top genes
genelist <- SelectMelocyteSkinSpecificGenes(Melfc=2, SkinFC=1.5)[[1]]
    genelist <- genelist[1:min(nrow(genelist),100),1]

#Add protein expression for top genes
  ProteinExpression <- function(genelist=genelist){  
      O1 <- SkinAtlasExpress(gene=as.data.frame(genelist)[,1])
      O1 <- O1[order(O1$Melanocyte, decreasing = T),]
    return(O1)
  }
  
  ProteinExpression(genelist=genelist)  


####PLot of expression

SkinPlot <- function(gene="MLANA"){
  require(ggplot2)
  require(cowplot)
  if(!exists("MeanCluster_Fantom_t")) {
    load("./Data/MembraneDataAll.RData")
  }
  df <- Fantom
  plotdata1 <- df[grep(gene, df$X),]
  MelanocyteIndex <- grep("Melan", names( plotdata1))
  MelanocyteMax <- apply( plotdata1[,MelanocyteIndex], 1, max)
  index <- which(MelanocyteMax==max(MelanocyteMax))
  plotdata1 <- data.frame(t(plotdata1[index,]))
  colnames(plotdata1) <- "Gene"
  plotdata1 <- data.frame(plotdata1[-1,])
  colnames(plotdata1) <- "Gene"
  plotdata1$Gene <- as.numeric(as.character(plotdata1$Gene))
  plotdata1$Tissue <- row.names(plotdata1)
  p <- ggplot(plotdata1, aes(y=Gene, x=Tissue))+geom_point() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
  #Gtex
  if(!exists("Gtex")) {
    load("./Data/MembraneDataAll.RData")
  }
  
  PlotDataGtex <-data.frame(Gtex[,c(1,3:55)])
  PlotDataGtex <- PlotDataGtex[PlotDataGtex$Description==gene,]
  PlotDataGtex <- PlotDataGtex[-1]
  PlotDataGtex <- data.frame(t(PlotDataGtex))
  PlotDataGtex$Tissue <- row.names(PlotDataGtex)
  colnames(PlotDataGtex)[1]<- "Gene"
  q <- ggplot(  PlotDataGtex, aes(y=Gene, x=Tissue))+geom_point() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ggtitle(gene)
  q
  
  SkinAtlasPlotData <- skinatlas[grep(gene, skinatlas$Gene.names),]
  if(nrow(SkinAtlasPlotData)==0){SkinAtlasPlotData = skinatlas[1,]
  SkinAtlasPlotData[1:13] <- 0
  SkinAtlasPlotData[14:16] <- "O"
  }
  SkinAtlasPlotData1 <- SkinAtlasPlotData[1:13] 
  SkinAtlasPlotData1 <- data.frame(t(  SkinAtlasPlotData1 ))
  colnames(SkinAtlasPlotData1) <- "Exprs"
  SkinAtlasPlotData1$Tissue <- row.names(SkinAtlasPlotData1)
  pp <- ggplot(  SkinAtlasPlotData1, aes(y=Exprs, x=Tissue))+geom_point() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ggtitle(SkinAtlasPlotData$Gene.names)
  pp
  
  print(plot_grid(p,plot_grid(q,pp, labels=c("GTEX", "Protein Expression")),nrow=2, labels="Fantom5"))
  #return(list(p,q,pp))
}

Output <- SkinPlot(gene="MLANA")
plot_grid(Output[[1]],plot_grid(plotlist = c(Output[[2]],Output[[3]])),nrow=2)

Output <- SkinPlot(gene="KRTAP19-1")

MarkerList <- c("HIBCH","ASPH","TYR",
                "CDH3","PLP1","KIAA1598","SOX10",
                "DAB2","SORBS1",#"TTYH3","TPCN2",
                "FYCO1","CHCHD6","LIMS2")

MarkerList2 <- c("MBP", "TNS3",
                 "LYST",#"PLEKHA5",
                 "TIMP2","GOLGA7B",
                 "SLC6A17",#"MED15",
                 #"MCF2L",
                 "SLC6A8",
                 #"HPS4",
                 "TSPAN10","CCL18","ACP5",
                 "PRKD3",#"RGS12",
                 "PDE3A","TNS1")

MarkerList <- SelectMelocyteSkinSpecificGenes(Melfc=10, SkinFC=0)[,1]
c(MarkerList)

pdf("MelanocyteMarkers3.pdf", width=15, height = 15)
for (i in c(MarkerList$genename)){print(i)
  SkinPlot(gene=i)
}

dev.off()


pdf("MelanocyteMarkers2.pdf", width=15, height = 15)
for (i in MarkerList2){
  SkinPlot(gene=i)
}
dev.off()

SkinPlot(gene="MBP")


#######Selection of Melanocyte Markers
#Step1: GTEX for skin selective expression
Markers1 <- SkinSelective()
Markers1 <- dim(Markers1[Markers1$SkinSelFC>2,]) #302

#Step2:
DFSkin <- SkinSelective(brain = TRUE) #Calculate fold change of Skin vs all other tissues in Gtex
DFMela <- MelanocyteSelective(num=1) #Calculate fold change of melanocyte vs all other cells in Fantom%

  DFList <- SelectMelocyteSkinSpecificGenes(Melfc=2, SkinFC=0)[[1]][,]
  MelaProtExpre <- ProteinExpression(genelist=DFList$Gene) [,c(14,15,12)]

  IndexList <- NULL
  OutPut <- NULL
  for (i in DFList$Gene){
    print(i)
    IndexList <- c(IndexList, grep(paste0("^",i,"$"), skinatlas$Entry.name))
    IndexList <- c(IndexList, grep(paste0("^",i,"$"), skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0(i,"$"),     skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0("^",i," "), skinatlas$Gene.names))
    IndexList <- c(IndexList, grep(paste0(" ",i," "), skinatlas$Gene.names))
    IndexList <- unique(IndexList)
    if(identical(IndexList, integer(0)))IndexList=0
    if(identical(IndexList, numeric(0)))IndexList=0
    print(IndexList)
    if(IndexList !=0){
        Ind <- which(skinatlas$Melanocyte[IndexList]==max(skinatlas$Melanocyte[IndexList]))
        IndexList <- IndexList[Ind]
    }
    y = if(IndexList !=0){ 
               skinatlas$Melanocyte[IndexList]
              }  else {NA}
    y1 = if(IndexList !=0){ 
      skinatlas$Entry.name[IndexList]
    }  else {NA}
    y2 = if(IndexList !=0){ 
      skinatlas$Gene.names[IndexList]
    }  else {NA}
    
  
  OutPut <- rbind(OutPut, data.frame(i, y1,y2,IndexList,y))
    IndexList <- NULL
  }
  
 
    
  Output <- data.frame(DFList, OutPut)
      Output[order( Output$MelaSelFC, decreasing = T),]
#SelectMelocyteSkinSpecificGenes()

      SkinPlot(gene="TTYH2")
View(Output)
